webapp/src/DOMJudgeBundle/Service/ScoreboardService.php

```diff
@@ -325,36 +322,61 @@ public function calculateScoreRow(Contest $contest, Team $team, Problem $problem
 
         foreach ($submissions as $submission) {
             /** @var Judging|null $judging */
-            $judging    = $submission->getJudgings()->first() ?: null;
-            $absSubmitTime = (float)$submission->getSubmittime();
-            $submitTime = $contest->getContestTime($absSubmitTime);
+            $judging = $submission->getJudgings()->first() ?: null;
 
             // Check if this submission has a publicly visible judging result:
             if ($judging === null || ($verificationRequired && !$judging->getVerified()) || empty($judging->getResult())) {
-                $pendingJury++;
+                // For the jury: only consider it pending if we don't have a correct one yet
+                // This is needed because during the freeze we consider submissions after the
+                // correct one for the public to not leak any info
+                if (!$correctJury) {
+                    $pendingJury++;
+                }
                 $pendingPubl++;
                 // Don't do any more counting for this submission.
                 continue;
             }
 
-            $submissionsJury++;
+            // We need to count the submission always, except when we don't want to count compiler
+            // penalties and the judging is a compiler error
+            $countSubmission = $compilePenalty || $judging->getResult() != Judging::RESULT_COMPILER_ERROR;
+
+            if (!$correctJury && $countSubmission) {
+                // For the jury: only consider it as a submission if we don't have a correct one yet
+                // This is needed because during the freeze we consider submissions after the
+                // correct one for the public to not leak any info
+                $submissionsJury++;
+            }
             if ($submission->isAfterFreeze()) {
-                // Show submissions after freeze as pending to the public (if SHOW_PENDING is enabled):
+                // Show submissions after freeze as pending to the public (if SHOW_PENDING is
+                // enabled). Note that we even show these submissions if they are a compiler-error
+                // and compile_penalty is set to false, to not leak any info
                 $pendingPubl++;
-            } else {
+            } elseif ($countSubmission) {
                 $submissionsPubl++;
             }
 
+            // If we encountered a correct submission during the whole contest, do not consider
+            // the submissions after that one for correctness
+            if ($correctJury) {
+                continue;
+            }
+
+            $absSubmitTime = (float)$submission->getSubmittime();
+            $submitTime    = $contest->getContestTime($absSubmitTime);
```

Comment:
$submission->getRelativeSubmittime?

Reply:
Do you think that is better than this? Imho doing it on the contest feels more logical, as it is relative to the contest start?

Reply:
A submission always belongs to exactly one contest and it removes one line from the code here, so I would be in favor of using the convenience function (or delete if we decide not to use it in general).

Reply:
Ah yeah didn't know the function was there already. I will use it.

Reply:
As discussed on Slack: we need both the relative time (to store in the scorecache) and the absolute time (to calculate first-to-solve) so we kept it this way.

We also still use the function in Submission, as that is used in the API. However, it was wrong (when the contest was paused or we had removed intervals), so I have fixed it (by making use of the function on Contest which already did the correct thing).